File: Week3/code/get_TreeHeight.py
#!/usr/bin/env py
"""
NAME
get_TreeHeight.py

DESCRIPTION
A script to calculate tree height use input file from command line.

FUNCTIONS
TreeHeight(degrees, distance)

DATA
No

VERSION
0.0.1

AUTHOR
Sirui Ye (sy1122@ic.ac.uk)
"""

import sys
import os
import numpy as np
import pandas as pd

# load file from command line (file path)
input_file = sys.argv[1]

# get file name
input_file_name = os.path.basename(input_file)

#strip off file extension
input_file_name = input_file_name.split(".")[0]

# load file data to dataframe
df = pd.read_csv(input_file)

def TreeHeight(degrees, distance):
    """[Calculate tree heights based on distance from tree base and angles from tree top]
       Args:
           degrees ([float])
           distance ([float])
       Returns:
           [list]: [heights of each tree]
       """
    
    radians = np.radians(degrees)
    height = distance * np.tan(radians)
    return height

# store tree heights in a new column in dataframe 
df["Tree.Height.m"] = TreeHeight(df["Angle.degrees"], df["Distance.m"])

# output file location
output_file = "../results/" + input_file_name + "_treeheights.csv"

# save output file 
df.to_csv(output_file,index=False) 

# compelte message 
print("All done.")

File: Week3/code/get_TreeHeight.R
#Author: Dashing Dingos
#Script: get_TreeHeight.R
#Created: Dec 2022
#Desc: Groupwork practical on tree height

# clear working directory
rm(list=ls())

# Function to calculate the height of a tree from its base angl to its top
TreeHeight <- function(degrees, distance) {
  radians <- degrees * pi / 180
  height <- distance * tan(radians)
  
  return (height)
}

# Import the data from the command line
# if the command line length is empty, use a default path
if (length( commandArgs(trailing = T) ) == 0 ){
  print("No files are inputted, using the default pathway. \n ../data/trees.csv ")
  fname = "../data/trees.csv"
} else {
  fname = commandArgs(trailing = T)[1]
}

# read the file as csv
data <- read.csv(file = fname, header = TRUE)

# make a new column on treesdata with the tree height results
data$TreeHeight.m <- TreeHeight(data$Angle.degrees, data$Distance.m)

# get the base file name without the file path or extension
fbase = tools::file_path_sans_ext(basename(fname))

# Write the filepath to save the results
writepath = paste("../results/", fbase, "_treeheights.csv", sep = "")

# save to results
print("Saving results into the results folder")
write.csv(data, writepath)

# Echo message when complete
print("All done in R!")